### ISSUE TITLE
score_one method modifies anomaly.LocalOutlierFactor internal state unintentionally?

### ISSUE THREAD

**Issue by MarekWadinger** [online-ml/river#1331]
## Versions

**river version**: development
**Python version**: Python 3.10.1
**Operating system**: macOS Sonoma 14.0 (23A344)

## Describe the bug

Hello 👋

I found out that `score_one` method of `anomaly.LocalOutlierFactor` changes internal state of the detector.

If this is not intentional I'd like to assist in resolving this issue.  I tried to trace back the error and found out that the problem is connected to `expand_objects` function, which creates only a references to original variables. Following modification to lines 349-369 in `anomaly.LocalOutlierFactor` [HERE](https://github.com/online-ml/river/blob/1a32e0d04eecbf79d8977c1a419d5b336d82bc77/river/anomaly/lof.py#L349), helped me to preserve the state over `score_one` calls.

```python
        (
            nm,
            x_list_copy,
            neighborhoods,
            rev_neighborhoods,
            k_dist,
            reach_dist,
            dist_dict,
            local_reach,
            lof,
        ) = expand_objects(
            self.x_scores,
            x_list_copy,
            self.neighborhoods,
            self.rev_neighborhoods,
            self.k_dist,
            self.reach_dist,
            self.dist_dict,
            self.local_reach,
            self.lof,
        )
```
to 
```python
        import copy
        (
            nm,
            x_list_copy,
            neighborhoods,
            rev_neighborhoods,
            k_dist,
            reach_dist,
            dist_dict,
            local_reach,
            lof,
        ) = expand_objects(
            self.x_scores,
            x_list_copy,
            self.neighborhoods.copy(),
            self.rev_neighborhoods.copy(),
            self.k_dist.copy(),
            copy.deepcopy(self.reach_dist),
            copy.deepcopy(self.dist_dict),
            self.local_reach.copy(),
            self.lof.copy(),
        )
```

Please, let me know if that makes sense, and I'd be happy to elaborate on any related issues or comments.

Thank you 🙏

## Steps/code to reproduce

```python
from river import anomaly

lof = anomaly.LocalOutlierFactor()

X = [{"a": 1, "b": 1}, {"a": 1, "b": 1}]
for x in X:
    lof.learn_one(x)

print(
    lof.x_list,
    lof.neighborhoods,
    lof.rev_neighborhoods,
    lof.k_dist,
    lof.reach_dist,
    lof.dist_dict,
    lof.local_reach,
    lof.lof, sep="\n"
    )

lof.score_one({"a": 0.5, "b": 1})

print(
    lof.x_list,
    lof.neighborhoods,
    lof.rev_neighborhoods,
    lof.k_dist,
    lof.reach_dist,
    lof.dist_dict,
    lof.local_reach,
    lof.lof, sep="\n"
    )
```
returns
```
[{'a': 1, 'b': 1}, {'a': 1, 'b': 1}]
{0: [1], 1: [0]}
{0: [1], 1: [0]}
{0: 0.0, 1: 0.0}
{0: {1: 0.0}, 1: {0: 0.0}}
{0: {1: 0.0}, 1: {0: 0.0}}
{0: 0, 1: 0}
{0: 0, 1: 0}

[{'a': 1, 'b': 1}, {'a': 1, 'b': 1}]
{0: [1, 2], 1: [0, 2], 2: [0, 1]}
{0: [1, 2], 1: [0, 2], 2: [0, 1]}
{0: 0.5, 1: 0.5, 2: 0.5}
{0: {1: 0.5, 2: 0.5}, 1: {0: 0.5, 2: 0.5}, 2: {0: 0.5, 1: 0.5}}
{0: {1: 0.0, 2: 0.5}, 1: {0: 0.0, 2: 0.5}, 2: {0: 0.5, 1: 0.5}}
{0: 2.0, 1: 2.0, 2: 2.0}
{0: 1.0, 1: 1.0, 2: 1.0}
```

**Comment by MaxHalford** [comment]
Yep go for it :)

**Comment by smastelini** [comment]
Closed via #1330.
